src/main/java/teammates/ui/webapi/UpdateFeedbackSessionAction.java

```diff
@@ -103,21 +108,93 @@ public JsonResult execute() throws InvalidHttpRequestBodyException {
         }
+
+        boolean notifyAboutDeadlines = getBooleanRequestParamValue(Const.ParamsNames.NOTIFY_ABOUT_DEADLINES);
+
+        List<EmailWrapper> emailsToSend = new ArrayList<>();
+        if (!studentDeadlines.equals(oldStudentDeadlines)) {
+            processDeadlineExtensions(courseId, feedbackSession, oldStudentDeadlines, studentDeadlines,
+                    false, notifyAboutDeadlines, emailsToSend);
+        }
+
+        if (!instructorDeadlines.equals(oldInstructorDeadlines)) {
+            processDeadlineExtensions(courseId, feedbackSession, oldInstructorDeadlines, instructorDeadlines,
+                    true, notifyAboutDeadlines, emailsToSend);
+        }
+
+        taskQueuer.scheduleEmailsForSending(emailsToSend);
+
+        return new JsonResult(new FeedbackSessionData(feedbackSession));
+    }
+
+    private void processDeadlineExtensions(String courseId, FeedbackSessionAttributes session,
+            Map<String, Instant> oldDeadlines, Map<String, Instant> newDeadlines,
+            boolean areInstructors, boolean notifyUsers, List<EmailWrapper> emailsToSend) {
+        // Revoke deadline extensions
+        Map<String, Instant> deadlinesToRevoke = new HashMap<>(oldDeadlines);
+        deadlinesToRevoke.keySet().removeAll(newDeadlines.keySet());
+
+        deadlinesToRevoke.keySet().forEach(email ->
+                logic.deleteDeadlineExtension(courseId, session.getFeedbackSessionName(), email, areInstructors));
+
+        // Create deadline extensions
+        Map<String, Instant> deadlinesToCreate = new HashMap<>(newDeadlines);
+        deadlinesToCreate.keySet().removeAll(oldDeadlines.keySet());
+
+        deadlinesToCreate.entrySet()
+                .stream()
+                .map(entry ->
+                    DeadlineExtensionAttributes
+                            .builder(courseId, session.getFeedbackSessionName(), entry.getKey(), areInstructors)
+                            .withEndTime(entry.getValue())
+                            .build())
+                .forEach(deadlineExtension -> {
+                    try {
+                        logic.createDeadlineExtension(deadlineExtension);
+                    } catch (InvalidParametersException | EntityAlreadyExistsException e) {
+                        log.severe("Unexpected error while creating deadline extension", e);
+                    }
+                });
+
+        // Update deadline extensions
+        Map<String, Instant> deadlinesToUpdate = new HashMap<>(newDeadlines);
+        deadlinesToUpdate = deadlinesToUpdate.entrySet().stream()
+                .filter(entry -> oldDeadlines.containsKey(entry.getKey())
+                        && !entry.getValue().equals(oldDeadlines.get(entry.getKey())))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        deadlinesToUpdate.entrySet()
+                .stream()
+                .map(entry -> DeadlineExtensionAttributes
```

Comment:
Line break is inconsistent with line 154. Can I check if this is on purpose?

Reply:
Sorry need to fix my IDE, it keeps changing between 4 and 8 spaces for some reason. I'll standardize the indentation.